core/include/gnuradio-4.0/Transactions.hpp

```diff
@@ -42,20 +42,46 @@ class CtxSettings : public SettingsBase {
      * The predicate will be called until it returns "true" (a match is found), or until it returns std::nullopt,
      * which indicates that no matches were found and there is no chance of matching anything in a further round.
      */
-    using MatchPredicate = std::function<std::optional<bool>(const property_map &, const property_map &, std::size_t)>;
-
-    TBlock                                       *_block = nullptr;
-    mutable std::mutex                            _lock{};
-    property_map                                  _active{};
-    property_map                                  _staged{};
-    std::set<std::string, std::less<>>            _auto_update{};
-    std::set<std::string, std::less<>>            _auto_forward{};
-    std::unordered_map<SettingsCtx, property_map> _settings{};
-    property_map                                  _default_settings{};
-    MatchPredicate                                _match_pred = nullMatchPred;
+    using MatchPredicate = std::function<std::optional<bool>(const pmtv::pmt &, const pmtv::pmt &, std::size_t)>;
+
+    // pmtv::pmt comparison is needed to use it as a key of std::map
+    struct PMTCompare {
+        bool
+        operator()(const pmtv::pmt &lhs, const pmtv::pmt &rhs) const {
+            // If the types are different, cast rhs to the type of lhs and compare
+            if (lhs.index() != rhs.index()) {
+                // TODO: throw if types are not the same?
+                return lhs.index() < rhs.index();
+            } else {
+                return std::visit(
+                        [&](const auto &left) -> bool {
+                            using T = std::decay_t<decltype(left)>;
+                            if constexpr ((pmtv::String<T> || pmtv::Scalar<T>) && !pmtv::Complex<T>) {
+                                return left < std::get<T>(rhs);
+                            } else {
+                                throw std::runtime_error("Invalid CtxSettings context type " + std::string(typeid(T).name()));
```

Comment:
replace with `gr::exception(...)`. 

We should aim at minimising throwing exceptions in the core lib core in favour of using `std::expected<..>`.